### PR TITLE
auth: add more info when connecting to auth

### PR DESCRIPTION
**Pull request by nkomonen-amazon**
This PR does the following:
- Requires a `source` argument for partner teams using the VS Code command `aws.codeWhisperer.connect`
  - We will use this to help identify which external extension is using the toolkit to connect
- In the metric `aws_loginWithBrowser` we are adding in the attribute `credentialStartUrl`
  - This will help us determine which type of SSO the user is using when they go through the browser login flow for SSO

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/codewhisperer/commands/basicCommands.ts [hunk 1]
```diff
@@ -143,48 +143,57 @@ export const showSsoSignIn = Commands.declare('aws.codeWhisperer.sso', () => asy
 // Shortcut command to directly connect to Identity Center or prompt start URL entry
 // It can optionally set a customization too based on given values to match on
 export const connectWithCustomization = Commands.declare(
-    'aws.codeWhisperer.connect',
```
Comment (nkomonen-amazon) [inline packages/core/src/codewhisperer/commands/basicCommands.ts:146]: This changed. The other changes are formatting related

File: packages/core/src/codewhisperer/commands/basicCommands.ts [hunk 2]
```diff
@@ -143,48 +143,57 @@ export const showSsoSignIn = Commands.declare('aws.codeWhisperer.sso', () => asy
...
-        //  * customizationNamePrefix: select customization by prefix, if `customizationArn` is `undefined`.
-        if (startUrl && region) {
-            await connectToEnterpriseSso(startUrl, region)
-        } else {
-            await getStartUrl()
-        }
+    { id: 'aws.codeWhisperer.connect', compositeKey: { 1: 'source' } },
+    () =>
+        async (
+            source: CodeWhispererSource,
```
Comment (nkomonen-amazon) [inline packages/core/src/codewhisperer/commands/basicCommands.ts:149]: This arg was added in. The other changes are formatting related.

File: packages/core/src/codewhisperer/commands/basicCommands.ts [hunk 3]
```diff
@@ -143,48 +143,60 @@ export const showSsoSignIn = Commands.declare('aws.codeWhisperer.sso', () => asy
...
+     * This command supports the following arguments:
+     * @param source - an identifier for who used this command. This value is not explicitly used in the function, but is used elsewhere.
+     * startUrl and region. If both arguments are provided they will be used, otherwise
+     *  the command prompts for them interactively.
+     * customizationArn: select customization by ARN. If provided, `customizationNamePrefix` is ignored.
+     * customizationNamePrefix: select customization by prefix, if `customizationArn` is `undefined`.
+     */
+    () =>
+        async (
+            source: string,
```
Comment (nkomonen-amazon) [inline packages/core/src/codewhisperer/commands/basicCommands.ts:157]: This was added in. The other changes are formatting related

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts [hunk 4]
```diff
@@ -199,16 +199,25 @@ export class SsoAccessTokenProvider {
...
+        const token = this.withBrowserLoginTelemetry(() => openBrowserAndWaitUntilComplete())
+
+        return this.formatToken(await token, registration)
+    }
+
+    /**
+     * Wraps the given function with telemetry related to the browser login.
+     */
+    private withBrowserLoginTelemetry<T extends (...args: any[]) => any>(func: T): ReturnType<T> {
+        if (telemetry.spans.some(s => s.name === 'aws_loginWithBrowser')) {
```
Comment (justinmk3) [inline packages/core/src/auth/sso/ssoAccessTokenProvider.ts:211]: If this is true, should we still call

    telemetry.record({ credentialStartUrl: this.profile.startUrl })

? That will update/ovewrite the `credentialStartUrl` field for the current span(s).

**Review by justinmk3 — APPROVED** [review]

File: packages/core/src/auth/sso/ssoAccessTokenProvider.ts (hunk 4 above)
Comment (nkomonen-amazon) [reply packages/core/src/auth/sso/ssoAccessTokenProvider.ts:211]: We're making the assumption that the parent `aws_loginWithBrowser` span (in `auth.ts`) has already set it.

I wanted to keep all values set in the parent one unless we had a reason not to.

**Pull request merged**